go-e Error

[main] INFO 2020/04/14 13:51:52 using config file /etc/evcc.yaml
[main] INFO 2020/04/14 13:51:52 listening at 0.0.0.0:7070
[go-e] WARN 2020/04/14 13:51:52 -- experimental --
[load] INFO 2020/04/14 13:51:52 main config: vehicle ✓ grid ✓ pv — charge —
[load] INFO 2020/04/14 13:51:52 main charge mode: off
[load] ERROR 2020/04/14 13:51:53 main charger error: invalid character '<' looking for beginning of value
[load] ERROR 2020/04/14 13:51:53 main charge controller error: invalid character '<' looking for beginning of value
[load] ERROR 2020/04/14 13:51:53 main charger error: invalid character '<' looking for beginning of value
panic: runtime error: index out of range [11] with length 0

Any ideas?

--- a/README.md
+++ b/README.md
@@ -7,12 +7,12 @@ EVCC is an extensible EV Charge Controller with PV integration implemented in [G
 ### Features
 
 - simple and clean user interface
-- multiple [chargers](#charger): Wallbe (tested with Wallbe Eco S), Phoenix controllers (similar to Wallbe), any other charger using scripting
-- more chargers experimentally supported: NRGKick, Go-E, SimpleEVSE, EVSEWifi
+- multiple [chargers](#charger): Wallbe (tested with Wallbe Eco S), Phoenix controllers (similar to Wallbe), go-eCharger, openWB slave, any other charger using scripting
+- more chargers experimentally supported: NRGKick, SimpleEVSE, EVSEWifi
 - different [vehicles](#vehicle) to show battery status: Audi (eTron), BMW (i3), Tesla, Nissan (Leaf), any other vehicle using scripting
 - integration with home automation - supports shell scripts and MQTT
 - status notifications using [Telegram](https://telegram.org) and [PushOver](https://pushover.net)
-- logging using [InfluxDB](https://www.influxdata.com)
+- logging using [InfluxDB](https://www.influxdata.com) and [Grafana](https://grafana.com/grafana/)
 - soft ramp-up/ramp-down of charge current ensures contactor only switched at minimum current
 - electric contactor protection
 - REST API

--- a/charger/go-e.go
+++ b/charger/go-e.go
@@ -15,14 +15,14 @@ const (
 
 // goeStatusResponse is the API response if status not OK
 type goeStatusResponse struct {
-	Car uint   `json:"car,string"` // car status
-	Alw uint   `json:"alw,string"` // allow charging
-	Amp uint   `json:"amp,string"` // current [A]
-	Err uint   `json:"err,string"` // error
-	Stp uint   `json:"stp,string"` // stop state
-	Tmp uint   `json:"tmp,string"` // temperature [°C]
-	Dws uint   `json:"dws,string"` // energy [Ws]
-	Nrg []uint `json:"nrg"`        // voltage, current, power
+	Car int   `json:"car,string"` // car status
+	Alw int   `json:"alw,string"` // allow charging
+	Amp int   `json:"amp,string"` // current [A]
+	Err int   `json:"err,string"` // error
+	Stp int   `json:"stp,string"` // stop state
+	Tmp int   `json:"tmp,string"` // temperature [°C]
+	Dws int   `json:"dws,string"` // energy [Ws]
+	Nrg []int `json:"nrg"`        // voltage, current, power
 }
 
 // GoE charger implementation
@@ -93,7 +93,7 @@ func (c *GoE) Enabled() (bool, error) {
 func (c *GoE) Enable(enable bool) error {
 	var status goeStatusResponse
 
-	var b uint
+	var b int
 	if enable {
 		b = 1
 	}
@@ -114,7 +114,7 @@ func (c *GoE) MaxCurrent(current int64) error {
 	uri := c.apiURL(goePayload) + fmt.Sprintf("amp=%d", current)
 
 	_, err := c.GetJSON(uri, &status)
-	if err == nil && status.Amp != uint(current) {
+	if err == nil && int64(status.Amp) != current {
 		return errors.New("update failed")
 	}
 

--- a/charger/go-e.go
+++ b/charger/go-e.go
@@ -1,7 +1,6 @@
 package charger
 
 import (
-	"errors"
 	"fmt"
 	"strings"
 
@@ -68,7 +67,7 @@ func (c *GoE) Status() (api.ChargeStatus, error) {
 	case 3, 4:
 		return api.StatusB, nil
 	default:
-		return api.StatusNone, fmt.Errorf("unknown result %d", status.Car)
+		return api.StatusNone, fmt.Errorf("car unknown result: %d", status.Car)
 	}
 }
 
@@ -85,7 +84,7 @@ func (c *GoE) Enabled() (bool, error) {
 	case 1:
 		return true, nil
 	default:
-		return false, fmt.Errorf("unknown result %d", status.Alw)
+		return false, fmt.Errorf("alw unknown result: %d", status.Alw)
 	}
 }
 
@@ -102,7 +101,7 @@ func (c *GoE) Enable(enable bool) error {
 
 	_, err := c.GetJSON(uri, &status)
 	if err == nil && status.Alw != b {
-		return errors.New("update failed")
+		return fmt.Errorf("alw update failed: %d", status.Amp)
 	}
 
 	return err
@@ -115,7 +114,7 @@ func (c *GoE) MaxCurrent(current int64) error {
 
 	_, err := c.GetJSON(uri, &status)
 	if err == nil && int64(status.Amp) != current {
-		return errors.New("update failed")
+		return fmt.Errorf("amp update failed: %d", status.Amp)
 	}
 
 	return err

--- a/charger/go-e.go
+++ b/charger/go-e.go
@@ -1,7 +1,7 @@
 package charger
 
 import (
-	"encoding/json"
+	"errors"
 	"fmt"
 	"strings"
 
@@ -9,25 +9,26 @@ import (
 )
 
 const (
-	goeStatus apiFunction = "status"
+	goeStatus  apiFunction = "status"
+	goePayload apiFunction = "mqtt?payload="
 )
 
 // goeStatusResponse is the API response if status not OK
 type goeStatusResponse struct {
-	Car uint   `yaml:"car"` // car status
-	Alw uint   `yaml:"alw"` // allow charging
-	Amp uint   `yaml:"amp"` // current [A]
-	Err uint   `yaml:"err"`
-	Stp uint   `yaml:"stp"`
-	Tmp uint   `yaml:"tmp"`
-	Dws uint   `yaml:"dws"` // energy [Ws]
-	Nrg []uint `yaml:"nrg"` // voltage, current, power
+	Car uint   `json:"car,string"` // car status
+	Alw uint   `json:"alw,string"` // allow charging
+	Amp uint   `json:"amp,string"` // current [A]
+	Err uint   `json:"err,string"` // error
+	Stp uint   `json:"stp,string"` // stop state
+	Tmp uint   `json:"tmp,string"` // temperature [°C]
+	Dws uint   `json:"dws,string"` // energy [Ws]
+	Nrg []uint `json:"nrg"`        // voltage, current, power
 }
 
 // GoE charger implementation
 type GoE struct {
 	*api.HTTPHelper
-	URI string
+	uri string
 }
 
 // NewGoEFromConfig creates a go-e charger from generic config
@@ -42,36 +43,20 @@ func NewGoEFromConfig(log *api.Logger, other map[string]interface{}) api.Charger
 func NewGoE(URI string) *GoE {
 	c := &GoE{
 		HTTPHelper: api.NewHTTPHelper(api.NewLogger("go-e")),
-		URI:        URI,
+		uri:        strings.TrimRight(URI, "/"),
 	}
-
-	c.HTTPHelper.Log.WARN.Println("-- experimental --")
 
 	return c
 }
 
 func (c *GoE) apiURL(api apiFunction) string {
-	return fmt.Sprintf("%s/%s", strings.TrimRight(c.URI, "/"), api)
-}
-
-func (c *GoE) getJSON(url string, result interface{}) error {
-	b, err := c.GetJSON(url, result)
-	if err != nil && len(b) > 0 {
-		var error goeStatusResponse
-		if err := json.Unmarshal(b, &error); err != nil {
-			return err
-		}
-
-		return fmt.Errorf("response code: %d", error.Err)
-	}
-
-	return err
+	return fmt.Sprintf("%s/%s", c.uri, api)
 }
 
 // Status implements the Charger.Status interface
 func (c *GoE) Status() (api.ChargeStatus, error) {
 	var status goeStatusResponse
-	if err := c.getJSON(c.apiURL(goeStatus), status); err != nil {
+	if _, err := c.GetJSON(c.apiURL(goeStatus), &status); err != nil {
 		return api.StatusNone, err
 	}
 
@@ -80,9 +65,7 @@ func (c *GoE) Status() (api.ChargeStatus, error) {
 		return api.StatusA, nil
 	case 2:
 		return api.StatusC, nil
-	case 3:
-		return api.StatusB, nil
-	case 4:
+	case 3, 4:
 		return api.StatusB, nil
 	default:
 		return api.StatusNone, fmt.Errorf("unknown result %d", status.Car)
@@ -92,7 +75,7 @@ func (c *GoE) Status() (api.ChargeStatus, error) {
 // Enabled implements the Charger.Enabled interface
 func (c *GoE) Enabled() (bool, error) {
 	var status goeStatusResponse
-	if err := c.getJSON(c.apiURL(goeStatus), status); err != nil {
+	if _, err := c.GetJSON(c.apiURL(goeStatus), &status); err != nil {
 		return false, err
 	}
 
@@ -110,35 +93,49 @@ func (c *GoE) Enabled() (bool, error) {
 func (c *GoE) Enable(enable bool) error {
 	var status goeStatusResponse
 
-	uri := c.apiURL(goeStatus) + "/mqtt?alw="
+	var b uint
 	if enable {
-		uri += "1"
-	} else {
-		uri += "0"
+		b = 1
 	}
 
-	return c.getJSON(uri, status)
+	uri := c.apiURL(goePayload) + fmt.Sprintf("alw=%d", b)
+
+	_, err := c.GetJSON(uri, &status)
+	if err == nil && status.Alw != b {
+		return errors.New("update failed")
+	}
+
+	return err
 }
 
 // MaxCurrent implements the Charger.MaxCurrent interface
 func (c *GoE) MaxCurrent(current int64) error {
 	var status goeStatusResponse
-	uri := fmt.Sprintf(c.apiURL(goeStatus)+"/mqtt?amp=%d", current)
-	return c.getJSON(uri, status)
+	uri := c.apiURL(goePayload) + fmt.Sprintf("amp=%d", current)
+
+	_, err := c.GetJSON(uri, &status)
+	if err == nil && status.Amp != uint(current) {
+		return errors.New("update failed")
+	}
+
+	return err
 }
 
 // CurrentPower implements the Meter interface.
 func (c *GoE) CurrentPower() (float64, error) {
 	var status goeStatusResponse
-	err := c.getJSON(c.apiURL(goeStatus), status)
-	power := float64(status.Nrg[11]) * 10
+	_, err := c.GetJSON(c.apiURL(goeStatus), &status)
+	var power float64
+	if len(status.Nrg) == 16 {
+		power = float64(status.Nrg[11]) * 10
+	}
 	return power, err
 }
 
-// ChargedEnergy implements the ChargeRater interface.
+// ChargedEnergy implements the ChargeRater interface
 func (c *GoE) ChargedEnergy() (float64, error) {
 	var status goeStatusResponse
-	err := c.getJSON(c.apiURL(goeStatus), status)
+	_, err := c.GetJSON(c.apiURL(goeStatus), &status)
 	energy := float64(status.Dws) / 3.6e6
 	return energy, err
 }

